pulser/simulation/simresults.py

```diff
@@ -14,63 +14,328 @@
             raise ValueError(
                 "`basis_name` must be 'ground-rydberg', 'digital' or 'all'."
                 )
         self._basis_name = basis_name
+        self._sim_times = sim_times
+        self._results: Union[list[Counter], list[qutip.Qobj]]
+
+    @property
+    @abstractmethod
+    def states(self) -> list[qutip.Qobj]:
+        """Lists states of the system at simulation times."""
+        pass
+
+    @abstractmethod
+    def get_state(self, t: float) -> qutip.Qobj:
+        """Returns the state of the system at time t."""
+        pass
+
+    @abstractmethod
+    def get_final_state(self) -> qutip.Qobj:
+        """Returns the final state of the system."""
+        pass
+
+    @abstractmethod
+    def _calc_weights(self, t_index: int) -> ArrayLike:
+        """Computes the bitstring probabilities for sampled states."""
+        pass
+
+    def expect(self, obs_list: Sequence[Union[qutip.Qobj, ArrayLike]]
+               ) -> list[Union[float, complex, ArrayLike]]:
+        """Returns the expectation values of operators in obs_list."""
+        if not isinstance(obs_list, (list, np.ndarray)):
+            raise TypeError("`obs_list` must be a list of operators.")
+
+        qobj_list = []
+        for obs in obs_list:
+            if not (isinstance(obs, np.ndarray)
+                    or isinstance(obs, qutip.Qobj)):
+                raise TypeError("Incompatible type of observable.")
+            if obs.shape != (2**self._size, 2**self._size):
+                raise ValueError("Incompatible shape of observable.")
+            qobj_list.append(qutip.Qobj(obs))
+
+        return cast(list, qutip.expect(qobj_list, self.states))
+
+    def sample_state(self, t: float, n_samples: int = 1000,
+                     t_tol: float = 1.e-3) -> Counter:
+        """Returns the result of multiple measurements at time t.
+
+        Args:
+            t (float): Time at which the state is sampled.
+            n_samples (int): Number of samples to return.
+            t_tol (float): Tolerance for the difference between t and
+                closest time.
+        """
+        t_index = self._get_index_from_time(t, t_tol)
+        dist = np.random.multinomial(n_samples, self._calc_weights(t_index))
+        return Counter({np.binary_repr(
+                        i, self._size): dist[i] for i in np.nonzero(dist)[0]})
+
+    def sample_final_state(self, n_samples: int = 1000) -> Counter:
+        """Returns the result of multiple measurements of the final state."""
+        return self.sample_state(self._sim_times[-1], n_samples)
+
+    def plot(self, op: qutip.Qobj, fmt: str = '', label: str = '') -> None:
+        """Plots the expectation value of a given operator op.
+
+        Args:
+            op (qutip.Qobj): Operator whose expectation value is wanted.
+            fmt (str): Curve plot format.
+            label (str): Axis label.
+        """
+        plt.plot(self._sim_times, self.expect([op])[0], fmt, label=label)
```

Comment:
Yes, because we allow for a list of observables, so here we choose the first one in the list since we only have one observable. I agree it's a bit weird.

Reply:
Yeah... But I understand the need